New dashboard query endpoints and consolidate Dashboard API/Public/Embed parameter resolution code

camsaul opened the pull request:
Part of https://github.com/metabase/metaboat/issues/144

This PR adds new `POST /api/dashboard/:id/card/:card-id/query` and `POST /api/dashboard/:id/card/:card-id/query/:export-format` endpoints. These new endpoints validate that the passed `:parameters` are allowed for the Dashboard and Card in question.

We already had similar endpoints for public and embedded dashboards, e.g. `GET /api/public/dashboard/:uuid/card/:card-id`. As part of this PR I unified the parameter handling and DashCard query execution used in the public and embed endpoints with the new code in `metabase.api.dashboard`. Previously the way we resolved params was not totally consistent, e.g. the `public` endpoints attempted to resolve parameters by either `:id` _or_ `:slug`, whereas other endpoints such as the chain filter endpoints exclusively resolve things by `:id`. This PR consolidates things so parameters are consistently always resolved by `:id`, except for embedding which still resolves by `:slug` due to the way embedding works. As part of this project I had to rework a bunch of public/embed tests that created Dashboards/Dashcards without and `:parameters` or `:parameter_mappings` respectively, which now would trigger a 'no matching parameter' error. I did a little code cleanup there while I was at it.

Overall this PR makes our parameter handling story a little less hairy. Parameter handling is still a little bit complicated, but at least it's complicated in more consistent ways now

camsaul committed:
Code cleanup

camsaul commented on src/metabase/api/embed.clj:
```diff
@@ -151,19 +152,19 @@
-                                                       :target s/Any
-                                                       :value  s/Any}])
-  "Adds `value` to parameters with `slug` matching a key in `merged-id->value` and removes parameters without a
+(s/defn ^:private apply-slug->value :- (s/maybe [{:slug   su/NonBlankString
+                                                  :type   s/Keyword
+                                                  :target s/Any
+                                                  :value  s/Any}])
+  "Adds `value` to parameters with `slug` matching a key in `merged-slug->value` and removes parameters without a
```
Minor correctness tweaks. These functions actually are passed a map of parameter _slug_ to value, so it's super misleading that they all have names like `id->value`... fixed that.

camsaul commented on src/metabase/api/embed.clj:
```diff
@@ -172,21 +173,23 @@
-          :let          [param (get param-id->param (:parameter_id param-mapping))]
-          :when         param]
-      (assoc param :target (:target param-mapping)))))
+(s/defn ^:private resolve-dashboard-parameters :- [dashboard-api/ParameterWithID]
+  "Given a `dashboard-id` and parameters map in the format `slug->value`, return a sequence of parameters with `:id`s
+  that can be passed to various functions in the `metabase.api.dashboard` namespace such as
+  `metabase.api.dashboard/run-query-for-dashcard-async`."
+  [dashboard-id :- su/IntGreaterThanZero
+   slug->value  :- {s/Any s/Any}]
+  (let [parameters (db/select-one-field :parameters Dashboard :id dashboard-id)
+        slug->id   (into {} (map (juxt :slug :id)) parameters)]
+    (vec (for [[slug value] slug->value
+               :let         [slug (u/qualified-name slug)]]
+           {:slug  slug
+            :id    (or (get slug->id slug)
+                       (throw (ex-info (tru "No matching parameter with slug {0}. Found: {1}" (pr-str slug) (pr-str (keys slug->id)))
+                                       {:status-code          400
+                                        :slug                 slug
+                                        :dashboard-parameters parameters})))
+            :value value}))))
```
This code is simplified a bit now and just maps the parameter slug to the correct parameter; the new stuff in `metabase.api.dashboard` can handle the DashCard parameter mapping stuff now.

camsaul commented on src/metabase/api/public.clj:
```diff
@@ -182,113 +184,29 @@
-      [{:card_id 1, :target [:variable [:template-tag :num]], :parameter_id \"537e37b4\"}]
-
-  Thus for public Dashboards JSON-encoded style we can look for a matching Dashcard parameter mapping, based on
-  `:target`, and then find the matching Dashboard parameter, based on `:id`.
-
-  *Cries*
-
-  TODO -- Tom has mentioned this, and he is very correct -- our lives would be much easier if we just used slug-style
-  for everything, rather than the weird JSON-encoded format we use for public Dashboards. We should fix this!"
-  [dashboard-id :- su/IntGreaterThanZero, query-params :- (s/maybe [{s/Keyword s/Any}])]
-  (when (seq query-params)
-    (let [dashboard-params        (db/select-one-field :parameters Dashboard, :id dashboard-id)
-          slug->dashboard-param   (u/key-by :slug dashboard-params)
-          dashcard-param-mappings (dashboard->dashcard-param-mappings dashboard-id)]
-      (for [{slug :slug, target :target, :as query-param} query-params
-            :let [target (normalize/normalize-tokens target :ignore-path)
-                  dashboard-param
-                  (or
-                   ;; try to match by slug...
-                   (slug->dashboard-param slug)
-                   ;; ...if that fails, try to find a DashboardCard param mapping with the same target...
-                   (matching-dashboard-param-with-target dashboard-params dashcard-param-mappings target)
-                   ;; ...but if we *still* couldn't find a match, throw an Exception, because we don't want people
-                   ;; trying to inject new params
-                   (throw (ex-info (tru "Invalid param: {0}" slug)
-                                   {:type qp.error-type/invalid-parameter})))]]
-        (merge query-param dashboard-param)))))
-
-(defn- check-card-is-in-dashboard
-  "Check that the Card with `card-id` is in Dashboard with `dashboard-id`, either in a DashboardCard at the top level or
-  as a series, or throw an Exception. If not such relationship exists this will throw a 404 Exception."
-  [card-id dashboard-id]
-  (api/check-404
-   (or (db/exists? DashboardCard
-         :dashboard_id dashboard-id
-         :card_id      card-id)
-       (when-let [dashcard-ids (db/select-ids DashboardCard :dashboard_id dashboard-id)]
-         (db/exists? DashboardCardSeries
-           :card_id          card-id
-           :dashboardcard_id [:in dashcard-ids])))))
```
All of this deleted code has been consolidated into the new unified Dashboard parameter handling code in `metabase.api.dashboard`

camsaul commented on test/metabase/api/card_test.clj:
```diff
@@ -139,7 +139,7 @@
                         {:database (u/the-id db)
                          :type     :native
                          :native   {:query         "SELECT COUNT(*) FROM VENUES WHERE CATEGORY_ID = {{category}};"
-                                    :template-tags {:category {:id           "a9001580-3bcc-b827-ce26-1dbc82429163"
```
These `:id` changes aren't meaningful but they make test output a little easier to look at.

camsaul commented on test/metabase/api/card_test.clj:
```diff
@@ -149,6 +149,21 @@
 (defmacro ^:private with-temp-native-card-with-params {:style/indent 1} [[db-binding card-binding] & body]
   `(do-with-temp-native-card-with-params (fn [~(or db-binding '_) ~(or card-binding '_)] ~@body)))
 
+(deftest run-query-with-parameters-test
+  (testing "POST /api/card/:id/query"
+    (testing "should respect `:parameters`"
+      (with-temp-native-card-with-params [{db-id :id} {card-id :id}]
+        (is (schema= {:database_id (s/eq db-id)
+                      :row_count   (s/eq 1)
+                      :data        {:rows     (s/eq [[8]])
+                                    s/Keyword s/Any}
+                      s/Keyword    s/Any}
+                     (mt/user-http-request
+                      :rasta :post 202 (format "card/%d/query" card-id)
+                      {:parameters [{:type   :category
+                                     :target [:variable [:template-tag :category]]
+                                     :value  2}]})))))))
```
This test is unrelated to the new stuff in this PR, but I couldn't find any similar tests in the namespace, so I went ahead and added it here.

camsaul commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -1481,3 +1483,134 @@
+                                         {:parameters {"_PRICE_" 3}}))))
+          (testing "Should ignore parameters that are valid for the Dashboard but not part of this Card (no mapping)"
+            (testing "Sanity check"
+              (is (schema= (dashboard-card-query-expected-results-schema :row-count 6)
+                           (mt/user-http-request :rasta :post 202 url
+                                                 {:parameters [{:id    "_PRICE_"
+                                                                :value 4}]}))))
+            (mt/with-temp-vals-in-db DashboardCard dashcard-id {:parameter_mappings []}
+              (is (schema= (dashboard-card-query-expected-results-schema :row-count 100)
+                           (mt/user-http-request :rasta :post 202 url
+                                                 {:parameters [{:id    "_PRICE_"
+                                                                :value 4}]})))))
+
+          ;; don't let people try to be sneaky and get around our validation by passing in a different `:target`
+          (testing "Should ignore incorrect `:target` passed in to API endpoint"
+            (is (schema= (dashboard-card-query-expected-results-schema :row-count 6)
+                         (mt/user-http-request :rasta :post 202 url
+                                               {:parameters [{:id     "_PRICE_"
+                                                              :target [:dimension [:field (mt/id :venues :id) nil]]
+                                                              :value  4}]})))))))))
+
+(defn- parse-export-format-results [^bytes results export-format]
+  (with-open [is (java.io.ByteArrayInputStream. results)]
+    (streaming.test-util/parse-result export-format is)))
+
+(deftest dashboard-card-query-export-format-test
+  (testing "POST /api/dashboard/:dashboard-id/card/:card-id/query/:export-format"
+    (with-chain-filter-fixtures [{{dashboard-id :id} :dashboard, {card-id :id} :card, {dashcard-id :id} :dashcard}]
+      (doseq [export-format [:csv :json :xlsx]]
+        (testing (format "Export format = %s" export-format)
+          (let [url (format "%s/%s" (dashboard-card-query-url dashboard-id card-id) (name export-format))]
+            (is (= (streaming.test-util/process-query-basic-streaming
+                    export-format
+                    (mt/mbql-query venues {:filter [:= $price 4]}))
+                   (parse-export-format-results
+                    (mt/user-http-request :rasta :post 200 url
+                                          {:request-options {:as :byte-array}}
+                                          :parameters (json/generate-string [{:id    "_PRICE_"
+                                                                              :value 4}]))
+                    export-format)))))))))
```
These are the actual tests for the new endpoints

camsaul commented on test/metabase/api/embed_test.clj:
```diff
@@ -47,28 +48,54 @@
+
+(defn do-with-temp-dashcard [{:keys [dash card dashcard]} f]
+  (with-temp-card [card card]
+    (mt/with-temp* [Dashboard     [dashboard (merge
+                                              (when-not (:parameters dash)
+                                                {:parameters [{:id      "_VENUE_ID_"
+                                                               :name    "Venue ID"
+                                                               :slug    "venue_id"
+                                                               :type    "id"
+                                                               :target  [:dimension (mt/id :venues :id)]
+                                                               :default nil}]})
+                                              dash)]
+                    DashboardCard [dashcard  (merge {:dashboard_id       (u/the-id dashboard)
+                                                     :card_id            (u/the-id card)
+                                                     :parameter_mappings (or (:parameter_mappings dashcard)
+                                                                             [{:parameter_id "_VENUE_ID_"
+                                                                               :card_id      (u/the-id card)
+                                                                               :target       [:dimension [:field (mt/id :venues :id) nil]]}])}
+                                                    dashcard)]]
+      (f dashcard))))
```
I had to rework this mock stuff here to create Dashboards that actually have real parameters. Otherwise the new unified param resolution code will complain that there's no matching parameter. That seems like a win anyway to me

camsaul commented on test/metabase/api/embed_test.clj:
```diff
@@ -257,58 +284,58 @@
 (deftest card-locked-params-test
   (with-embedding-enabled-and-new-secret-key
-    (with-temp-card [card {:enable_embedding true, :embedding_params {:abc "locked"}}]
+    (with-temp-card [card {:enable_embedding true, :embedding_params {:venue_id "locked"}}]
```
Since the tests now use *real* parameters I switched them from `:abc` (totally fake) to `:venue_id` which actually makes sense given the query for the mock Card created by `with-temp-card`

camsaul commented on test/metabase/query_processor/streaming/test_util.clj:
```diff
@@ -0,0 +1,107 @@
+(ns metabase.query-processor.streaming.test-util
+  "Utility functions for testing QP streaming (download) functionality."
+  (:require [cheshire.core :as json]
+            [clojure.data.csv :as csv]
+            [clojure.test :refer :all]
+            [dk.ative.docjure.spreadsheet :as spreadsheet]
+            [metabase.query-processor :as qp]
+            [metabase.query-processor.streaming :as qp.streaming]
+            [metabase.test :as mt])
+  (:import [java.io BufferedInputStream BufferedOutputStream ByteArrayInputStream ByteArrayOutputStream InputStream InputStreamReader]))
```
The stuff in this namespace used to be in `metabase.query-processor.streaming-test`, but I'm reusing some of it for testing exports in `metabase.api.dashboard-test`. That caused a circular reference, so I moved it into its own namespace.

camsaul committed:
Make the linters happy

camsaul committed:
Add pivot version of the new endpoints

dpsutton commented on src/metabase/api/dashboard.clj:
```diff
@@ -643,4 +648,126 @@
+              ;; TODO -- passing this `:middleware` map is a little repetitive, need to think of a way to not have to specify
+              ;; this all over the codebase any time we want to do a query with an export format. Maybe this should be the
+              ;; default if `export-format` isn't `:api`?
+              :middleware    {:process-viz-settings?  true
```
what if we had middleware groups? you can specify middleware and middleware groups and those would get expanded correctly?

camsaul replied:
that could be a good idea. Might make sense as a follow-on refactor

dpsutton commented on src/metabase/api/public.clj:
```diff
@@ -97,6 +94,19 @@
+    (qp.streaming/streaming-response
+        [{:keys [reducedf], :as context} export-format (u/slugify (:card-name info))]
+        (let [context  (assoc context :reducedf (public-reducedf reducedf))
+              in-chan  (binding [api/*current-user-permissions-set* (atom #{"/"})]
```
this binding makes me nervous. Does that permission set get conveyed onto the thread that does the work and populates the resulting channel? I'm guessing you are sure but just want to highlight it.

camsaul replied:
yeah, the binding does get picked up. It is a little bit of an awkward way of setting the perms but I'm not sure what a better way of doing it would be. Worst case scenario it would just fail closed.

dpsutton commented on src/metabase/api/public.clj:
```diff
@@ -113,15 +123,7 @@
     (m/mapply card-api/run-query-for-card-async card-id export-format
```
do we need this `m/apply` here?

camsaul replied:
yeah, if you look down a couple of lines we pass an `options` map as the last parameter

dpsutton commented:
Looks like the e2e tests are meaningful failures.

daltojohnso committed:
implement usage of the new dashboard card query endpoint (#19012)

* add new endpoints to services.js

* replace CardApi.query with the new endpoint + pass dashboardId

* add parameter id to parameter object found on query body

* run dashchards using card query endpoint when they're new/unsaved on dashboard

* Update endpoint references in e2e tests

* Remove count check from e2e test

We can make sure a double-mapped parameter filter shows results from
both fields, but unfortunately with the new endpoint, the results don't
seem to work anymore. I think that's OK? Maybe?

* skip corrupted filters test

the query endpoint now results in a 500 error caused by a schema
mismatch

* fix a few e2e intercepted request mismatches

* unskip filter corruption test and remove the part that no longer works
